enhancements/machine-config/auth-token.md

```diff
@@ -0,0 +1,124 @@
+---
+title: mco-auth-token
+authors:
+  - "@cgwalters"
+reviewers:
+  - "@crawford"
+approvers:
+  - "@ashcrow"
+  - "@crawford"
+  - "@imcleod"
+  - "@runcom"
+creation-date: 2020-08-05
+last-updated: 2020-08-19
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# Support a provisioning token for the Machine Config Server
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement proposes that for user-provisioned installations we optionally support generating a "provisioning token":
+
+```
+$ dd if=/dev/urandom bs=32 count=1 | base64 > token
+$ oc -n openshift-config create secret generic machineconfig-auth --from-file=token
+```
+
+Installer-provisioned infrastructure environments would implement this automatically.
+
+When configured, requests to the Machine Config Server will require this token to be provided via e.g. `GET /config/worker&token=<value>`.
+
+## Motivation
+
+See https://github.com/openshift/machine-config-operator/pull/784
+
+The Ignition configuration can contain secrets, and we want to avoid it being accessible both inside and outside the cluster.
+
+### Goals
+
+- Increase security
+- Avoid breaking upgrades
+- Easy to implement
+- Easy to understand for operators
+- Eventually support being replaced by something more sophisticated (e.g. TPM2 attestation)
+- Avoid too strong dependence on platform-specific functionality (i.e. UPI AWS should remain close to  UPI bare metal)
+
+### Non-Goals
+
+- Completely rework major parts of the provisioning process today
+
+## Proposal
+
+Change the MachineConfigServer to support requiring a token in order to retrieve the Ignition config.
+
+In IPI/machineAPI managed clusters, the MCO would automatically handle this end to end.
+
+### Implementation Details/Notes/Constraints
+
+The documentation for user provisioned infrastructure would be extended to show how to configure this.  We would also need to debate important details such as whether this is something we want to allow to be configured via `install-config.yaml` or whether it should only be a "day 2" hardening.
+
+For installer-provisioned infrastructure, this will be much easier to implement for installations where the `user-data` provided to nodes is owned by the MCO: https://github.com/openshift/enhancements/blob/master/enhancements/machine-config/user-data-secret-managed.md
+
+In order to avoid race conditions, the MCS might support the "previous" token for a limited period of time (e.g. one hour/day).
```

Comment:
Rotating the token will have implications for requirements for the ability to stage bare metal nodes before delivering them to remote sites for telco customers. If the worker node is installed on day 1, then it takes 3 days for it to be delivered to the site of a tower where it is powered back on and connected to the cluster, the token it has will have expired.

Reply:
Thanks for looking at this!

The short answer would be "don't rotate the token".

A better fix needs a lot more information on exactly what "stage" means.  I think a generally good workflow we want to support in CoreOS is using `coreos-installer` in this "staging" site and configuring networking etc. and have the "pointer ignition config" ready to fetch the remote target.  In this scenario, the staging site would definitely need to use the correct token in the pointer config.  

I think the answer is probably just "don't rotate the token while any nodes are being staged".

Particularly for bare metal as I mention at the end, I think the clearly best solution is using TPM2 (e.g. in this case the "staging site" can communicate the hardware identity of the machine to the target in advance even), but that's much more involved.

Reply:
> Thanks for looking at this!
> 
> The short answer would be "don't rotate the token".

I think that means there would be no IPI for bare metal with remote workers, which would defeat the goal.

> A better fix needs a lot more information on exactly what "stage" means. I think a generally good workflow we want to support in CoreOS is using `coreos-installer` in this "staging" site and configuring networking etc. and have the "pointer ignition config" ready to fetch the remote target. In this scenario, the staging site would definitely need to use the correct token in the pointer config.

I'm not sure what the policy about linking to product requirement tickets in GitHub reviews is, so I'll try to summarize. We need to install the host at one location so that it is configured to join its cluster properly (that may mean the pointer config, or something else, it's unclear). Then some of the hosts will be physically moved to different remote sites, plugged into the network, and they need to rejoin the cluster without any user intervention (we can't assume the person installing the hardware is qualified to do more than rack it and plug it in).

IPI for bare metal does not use coreos-installer, it writes a disk image to a block device, including the pointer ignition configuration. See https://github.com/openshift/enhancements/blob/master/enhancements/baremetal/baremetal.md

It sounds like you're suggesting that during staging the image and ignition pointer should be written to the host, but the host shouldn't be booted until it is at the remote location to preserve the pointer configuration. Is that right? I don't expect that to be acceptable given the usual burn-in and validation requirements for these sorts of use cases.

> I think the answer is probably just "don't rotate the token while any nodes are being staged".

Is the lifetime for the automated rotation configurable?

What happens if the token does rotate after a host leaves the staging site and before it is connected back to the cluster? Does the user ship the host back to the staging site to be re-imaged?

> Particularly for bare metal as I mention at the end, I think the clearly best solution is using TPM2 (e.g. in this case the "staging site" can communicate the hardware identity of the machine to the target in advance even), but that's much more involved.

Reply:
OK I had missed you're talking about IPI metal, not UPI.  I think we probably need a whole section on this - the core problem here is that the "automatic IPI" case hinges on the MCO managing the user data, but it's not clear to me that applies in IPI/machineAPI metal (does it?).

In other words, forget about staging for a second - would this enhancement work as described in IPI metal or not?

Then once we have that part figured out...we can think about how we handle rotation.

It might be that for IPI metal, we don't rotate by default - or we add an `install-config.yaml` option to disable automatic rotation.

Reply:
> In other words, forget about staging for a second - would this enhancement work as described in IPI metal or not?

To rephrase; does https://github.com/openshift/enhancements/blob/master/enhancements/machine-config/user-data-secret-managed.md apply in IPI metal today?

Reply:
Metal IPI uses the user-data secret to pass the ignition pointer configuration to the host via cloudinit. So, yes, I think the user-data-secret-managed enhancement applies to metal IPI as well.

Reply:
For metal-ipi we also need to consider the issue of how to access the MCO managed config from the cluster-api-provider ref https://github.com/openshift/machine-config-operator/issues/1690 - today we only provide the pointer config via the config-drive, but that means a chicken/egg issue with advanced network configurations.

With this proposal, would we expect a privileged in-cluster controller to be able to access the MCS, or would we instead have to retrieve the underlying MachineConfig directly?

Reply:
I don't think this proposal makes hacking things absent https://github.com/openshift/machine-config-operator/issues/1690 significantly more difficult - a privileged pod could get the token to talk to the MCS too.

(It doesn't help either of course)